haxe: add 4.2.1 and update default version

sternenseemann opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

#116302 with a few cleanups and without removing the old compilers.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

locallycompact committed:
neko: 2.2.0 -> 2.3.0

Co-authored-by: sternenseemann <0rpkxez4ksa01gb3typccl0i@systemli.org>

sternenseemann committed:
haxe: 3.4.6 -> 4.2.1

haxe_4_2: init at 4.2.1

haxe_3_4, haxe_3_2 still exist.

Co-authored-by: Daniel Firth <dan.firth@homotopic.tech>

sternenseemann commented:
@ofborg build neko haxe haxe_3_2 haxe_3_4 haxe_4_0 haxe_4_1

sternenseemann commented:
@ofborg build neko haxe haxe_3_2 haxe_3_4 haxe_4_0 haxe_4_1

locallycompact commented:
Looks good to me.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 116953` at bf0dbb41 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built successfully:</summary>
  <ul>
    <li><a href="https://gist.github.com/4da64c2c971f41b016b9163cec8f031c">haxe</a></li> (haxe_4_2)</li>
    <li><a href="https://gist.github.com/10a29f9d55082e72ed052d8ac526b3d7">hxcpp</a></li> (haxePackages.hxcpp)</li>
    <li><a href="https://gist.github.com/2c173188bc3082e760c0bf5bfee123c3">haxePackages.hxcs</a></li></li>
    <li><a href="https://gist.github.com/4214d0f62b36a8632bbf0bc4832e0ed7">haxePackages.hxjava</a></li></li>
    <li><a href="https://gist.github.com/370e27b62b24026eb35d982ca3f16434">haxePackages.hxnodejs_4</a></li></li>
    <li><a href="https://gist.github.com/a9a2cf79142b40cc7d61bf3e1c7f7d30">haxe_3_2</a></li></li>
    <li><a href="https://gist.github.com/39b239117a608f2a911c61398ddd049d">haxe_3_4</a></li></li>
    <li><a href="https://gist.github.com/31e32f669ab3b460970dee5194730ef6">haxe_4_0</a></li></li>
    <li><a href="https://gist.github.com/9817bf6ff87823157d7cd89c0021c89d">haxe_4_1</a></li></li>
    <li><a href="https://gist.github.com/0cdf91bc25774ee6ce495996ae4e8e78">neko</a></li></li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/compilers/haxe/default.nix:51:7:
```
   |
51 |       installPhase = ''
   |       ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/haxe/default.nix:107:9:
```
    |
107 |         license = with licenses; [ gpl2 bsd2 /*?*/ ];  # -> docs/license.txt
    |         ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/neko/default.nix:36:5:
```
   |
36 |     license = licenses.lgpl21;
   |     ^
```
</li>
  </ul>
</details>

aszlig reviewed (approved):
LGTM, although I didn't take a *detailed* look at it, nor did I test it.

locallycompact commented:
@sternenseemann If you're interested I could use help with hashlink as well. https://github.com/NixOS/nixpkgs/pull/116321

zseri commented on pkgs/development/compilers/haxe/default.nix:
```diff
@@ -1,14 +1,39 @@
+    if lib.versionAtLeast version "4.0"
+    then with ocaml-ng.ocamlPackages; [
+      ocaml
+      findlib
+      sedlex_2
+      xml-light
+      xml-light
+      ptmap
+      camlp5
+      sha
+      dune_2
+      luv
```
```suggestion
      camlp5
      dune_2
      findlib
      luv
      ptmap
      sedlex_2
      sha
      xml-light
```

sternenseemann replied:
Alphabetical sorting doesn't matter.

zseri replied:
why is `xml-light` duplicated in the list?

sternenseemann replied:
Oversight. Thanks.

zseri replied:
Where does the sorting come from?

sternenseemann replied:
There is none and sorting is also generally not a good idea: If it were to make sense, we would need to enforce it. Enforcing sorted argument / input lists in nixpkgs would lead to more merge conflicts than the trouble would be worth.

Even in files which are supposed to be alphabetically sorted, like `all-packages.nix` etc. this rule is not enforced because it is not feasible.

zseri commented on pkgs/development/compilers/haxe/default.nix:
```diff
@@ -17,7 +42,9 @@ let
         rev = "refs/tags/${version}";
```
afaik `refs/tags/` shouldn't be needed.
```suggestion
        rev = version;
```

zseri replied:
what was the result here?

sternenseemann replied:
84bbcc9ba18

zseri commented on pkgs/development/compilers/haxe/default.nix:
```diff
@@ -17,7 +42,9 @@ let
 
-      inherit prePatch;
+      prePatch = args.prePatch or (''
+        sed -i -e 's|"neko"|"${neko}/bin/neko"|g' extra/haxelib_src/src/haxelib/client/Main.hx
```
```suggestion
        substituteInPlace extra/haxelib_src/src/haxelib/client/Main.hx \
          --replace '"neko"' '"${neko}/bin/neko"'
```

zseri commented on pkgs/development/compilers/haxe/default.nix:
```diff
@@ -78,7 +105,7 @@ let
         homepage = "https://haxe.org";
         license = with licenses; [ gpl2 bsd2 /*?*/ ];  # -> docs/license.txt
-        maintainers = [ maintainers.marcweber ];
+        maintainers = [ maintainers.marcweber maintainers.locallycompact ];
```
```suggestion
        maintainers = with maintainers; [ marcweber locallycompact ];
```

zseri replied:
Huh?

sternenseemann replied:
This `with` bindings are a matter of taste, really, and change nothing semantically. Also not changing this keeps the diff more readable and saves me the trouble of rebasing again.

zseri commented on pkgs/development/compilers/neko/default.nix:
```diff
@@ -32,7 +34,7 @@ stdenv.mkDerivation rec {
     homepage = "https://nekovm.org";
     license = licenses.lgpl21;
-    maintainers = [ maintainers.marcweber ];
+    maintainers = [ maintainers.marcweber maintainers.locallycompact ];
```
```suggestion
    maintainers = with maintainers; [ marcweber locallycompact ];
```

zseri commented on pkgs/development/ocaml-modules/extlib/1.7.8.nix:
```diff
@@ -0,0 +1,30 @@
+{ stdenv, lib, fetchurl, fetchpatch, ocaml, findlib, cppo, minimal ? true }:
+
+assert lib.versionAtLeast (lib.getVersion ocaml) "3.11";
+
```
```nix
let
  minimal_flag = if minimal then "minimal=1" else "";
in
```

zseri commented on pkgs/development/ocaml-modules/extlib/1.7.8.nix:
```diff
@@ -0,0 +1,30 @@
+
+  dontConfigure = true;      # Skip configure
+  # De facto, option minimal=1 seems to be the default.  See the README.
+  buildPhase     = "make ${if minimal then "minimal=1" else ""} build";
+  installPhase   = "make ${if minimal then "minimal=1" else ""} install";
```
```suggestion
  buildPhase     = "make ${minimal_flag} build";
  installPhase   = "make ${minimal_flag} install";
```

sternenseemann replied:
Probably not worth it refactoring this as 1.7.8 is only meant as a temporary thing in nixpkgs to update haxe. I promise I'll refactor the normal extlib expression when 1.7.9 comes around and this can be finally deleted :)

SuperSandro2000 commented on pkgs/development/compilers/haxe/default.nix:
```diff
@@ -1,14 +1,39 @@
+        ++ lib.optional (lib.versionAtLeast version "4.1") [ mbedtls ]
+        ++ ocamlDependencies version;
 
       src = fetchgit {
```
Can we use fetchFromGitHub?

sternenseemann replied:
I guess, but `fetchFromGitHub` also just calls `fetchgit` in that case.

SuperSandro2000 commented on pkgs/development/compilers/neko/default.nix:
```diff
@@ -32,7 +34,7 @@ stdenv.mkDerivation rec {
     description = "A high-level dynamically typed programming language";
     homepage = "https://nekovm.org";
     license = licenses.lgpl21;
```
lgpl21Only or lgpl21Plus?

SuperSandro2000 commented on pkgs/development/ocaml-modules/extlib/1.7.8.nix:
```diff
@@ -0,0 +1,30 @@
+  meta = {
+    homepage = "https://github.com/ygrek/ocaml-extlib";
+    description = "Enhancements to the OCaml Standard Library modules";
+    license = lib.licenses.lgpl21;
```
lgpl21Only or lpgl21Plus?

sternenseemann committed:
neko: try to list applying licenses more accurately

As a side note: This change shows why `with` can be dangerous business:
It doesn't shadow any existing bindings which can be unexpected. If I
were to use with licenses; [ … ] here, zlib in the with block would
actually be the zlib passed via the function arguments instead of the
zlib from licenses which would be expected. This was what caused the
previous eval error.

sternenseemann committed:
haxe*: set licenses according to upstream opam file

Upstream specifies MIT and GPL2+ in its opam file, so we run with this.
There doesn't seem to have been any license change and I couldn't track
down the mentioned docs/license.txt.

sternenseemann commented:
@ofborg build neko haxe haxe_3_2 haxe_3_4 haxe_4_0 haxe_4_1

sternenseemann commented:
@ofborg build neko haxe haxe_3_2 haxe_3_4 haxe_4_0 haxe_4_1

sternenseemann commented:
@ofborg build neko haxe haxe_3_2 haxe_3_4 haxe_4_0 haxe_4_1

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 116953` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
10 packages built:
</summary>
<ul>
<li>haxe (haxe_4_2)</li>
<li>hxcpp (haxePackages.hxcpp)</li>
<li>haxePackages.hxcs</li>
<li>haxePackages.hxjava</li>
<li>haxePackages.hxnodejs_4</li>
<li>haxe_3_2</li>
<li>haxe_3_4</li>
<li>haxe_4_0</li>
<li>haxe_4_1</li>
<li>neko</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
haxe:
<details>


warning: unclear-gpl
`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/haxe/default.nix:107:9:

```
    |
107 |         license = with licenses; [ gpl2 bsd2 /*?*/ ];  # -> docs/license.txt
    |         ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
</details>
</details>

sternenseemann commented:
@ofborg build neko haxe haxe_3_2 haxe_3_4 haxe_4_0 haxe_4_1

kvtb commented on pkgs/development/compilers/haxe/default.nix:
```diff
@@ -1,14 +1,38 @@
+    ];
+
+  # extraPatch gets appended to the default prePatch. If prePatch is given,
+  # it overrides the default prePatch and extraPatch is irrelevant
```
I'd keep it simple as in was.
Yes, some code was duplicated, but it did not require additional comment (with extraMentalLoad to a code reader)

kvtb replied:
I mean in the original code and here https://github.com/NixOS/nixpkgs/issues/95638#issuecomment-776875554

Also, Haxe 4 is not yet well tested on NixOS and may require more patches.

kvtb replied:
I used only NodeJs target

sternenseemann replied:
Yeah, more testing would be great! I also suspect haxe 4.x needs one or two extra seds like the 3.x versions.

idk, there is only one special case so its worth it I'd say, especially if we'll remove 3.2 at some point.

kvtb replied:
I meant all the `extraPatch` introduced for no profit, only `prePatch` was already enough

sternenseemann replied:
Since we only really needed `extraPatch` in 3.4, doesn't seem that worth it actually. Removed in favor of a simpler model.

kvtb commented on pkgs/development/compilers/haxe/default.nix:
```diff
@@ -1,14 +1,38 @@
+      sha
+      dune_2
+      luv
+    ] ++ (if lib.versionAtLeast version "4.2" then [
```
do you really want to maintain those `if`/`then` when next Haxe version is out ?
inlining `ocamlDependencies` function removes them

sternenseemann replied:
Sure, why not. I agree it's a bit ugly, but when extlib 1.7.9 comes out we'll be able to get rid of this.

kvtb replied:
I can predict what will be: at some point you will be so unhappy with similar contributions by others that you fork out and freeze `nixpkgs` for personal use (probably undoing there most of own contributions), and no one will maintain the ruins in `master`.

sternenseemann replied:
Not sure what you getting at, I generally maintain my contributions and have no intention of maintaining a fork of nixpkgs because why would I?

My intention with this function is merely to reduce code duplication, so we don't need to make the same adjustment in 10 places in the future if for example something in `ocamlPackages` changes.

sternenseemann replied:
Rebasing onto #117113 allows us to get rid of the extlib conditional already today.

sternenseemann commented:
I've dropped `haxe` 4.1 and 4.0 because they require us to jump through hoops for extlib as they need < 1.7.7. I can add them back if we absolutely need them, but would prefer not to.

OPNA2608 commented:
Built [lukaskj/haxe-example](https://github.com/lukaskj/haxe-example) targets "cpp", "cs", "java" and "executable" (neko) on Neko 4.2. All build results ran fine, either on their own or under their respective runtime.

Locally packaged [haxeui/hxWidgets](https://github.com/haxeui/hxWidgets) + dependencies, built one of its sample GUI projects on Neko 4.2 and 3.4 (this PR) and 3.4 (current master). All build results ran fine.

<details>
  <summary>(Used overlay for & screenshot of hxWidgets here)</summary>
  
```nix
self: super: rec {

  haxePackages = super.haxePackages // {
    hscript = super.haxePackages.buildHaxeLib rec {
      libname = "hscript";
      version = "2.4.0";
      sha256 = "0qdxgqb75j1v125l9xavs1d32wwzi60rhfymngdhjqhdvq72bhxx";
      meta.description = "Haxe Script is a scripting engine for a subset of the Haxe language";
    };
    hxWidgets = super.haxePackages.buildHaxeLib rec {
      libname = "hxWidgets";
      version = "1.7.2";
      sha256 = "1ki90m2qqyqmcmwd97id4ck9rr25dvdlnmfp53bg8prcyg41wbgf";
      meta.description = "Haxe externs (and wrappers) for wxWidgets";
      propagatedBuildInputs = with super; [ wxGTK31 libGL ];
    };
  };

  haxetest = super.callPackage (

{ stdenv
, haxe, haxePackages
}:

stdenv.mkDerivation rec {
  pname = "haxetest";
  version = "irrelevant";
  
  src = haxePackages.hxWidgets.src;
  
  nativeBuildInputs = [ haxe ] ++ (with haxePackages; [ hxcpp hscript hxWidgets ]);
  
  buildPhase = "cd samples/00-Showcase && haxe build.hxml";
  
  installPhase = ''
    install -Dm755 bin/Main $out/bin/00-Showcase
  '';
}
 
  ) { };
}
```

![grafik](https://user-images.githubusercontent.com/23431373/112451410-c1514f00-8d55-11eb-8dd7-496f16774137.png)

</details>

LGTM.

sternenseemann commented:
Great, thank you for your help! If there aren't any other comments, I think we can merge and then (re)introduce 4.0 and 4.1 in the next PR!

OPNA2608 commented:
As a more complex test, I also managed to get Friday Night Funkin' compiled & running with Haxe 4.2, though only v0.2.7.1. Building the current master with some required patches gets… *stuck?* Haxelib, Neko and its GCC subprocesses just keep eating away at the CPU without any new compiled results. I assume its a problem with one of the dependencies/FNF and not Haxe/Neko itself.

![grafik](https://user-images.githubusercontent.com/23431373/112506604-38a2d500-8d8e-11eb-87c6-af824a8241ec.png)

I'll still look forward to the 4.1 PR :slightly_smiling_face:

sternenseemann commented:
@ofborg build haxe_3_2 haxe_3_4

sternenseemann commented:
@ofborg build haxe_3_2 haxe_3_4 neko
